web/server/web_client.c

```diff
@@ -812,109 +849,348 @@ static inline char *http_header_parse(struct web_client *w, char *s, int parse_u
     return ve;
 }
 
-// http_request_validate()
-// returns:
-// = 0 : all good, process the request
-// > 0 : request is not supported
-// < 0 : request is incomplete - wait for more data
-
-typedef enum {
-    HTTP_VALIDATION_OK,
-    HTTP_VALIDATION_NOT_SUPPORTED,
-#ifdef ENABLE_HTTPS
-    HTTP_VALIDATION_INCOMPLETE,
-    HTTP_VALIDATION_REDIRECT
-#else
-    HTTP_VALIDATION_INCOMPLETE
-#endif
-} HTTP_VALIDATION;
-
-static inline HTTP_VALIDATION http_request_validate(struct web_client *w) {
-    char *s = (char *)buffer_tostring(w->response.data), *encoded_url = NULL;
+static inline HTTP_VALIDATION web_client_is_complete(char *begin,char *end,size_t length){
+    if ( begin == end){
+        return HTTP_VALIDATION_INCOMPLETE;
+    }
 
-    size_t last_pos = w->header_parse_last_size;
-    if(last_pos > 4) last_pos -= 4; // allow searching for \r\n\r\n
-    else last_pos = 0;
+    if ( length > 3  ){
+        begin = end - 4;
+    }
 
-    w->header_parse_tries++;
-    w->header_parse_last_size = buffer_strlen(w->response.data);
+    uint32_t counter = 0;
+    do{
+        if (*begin == '\r'){
+            begin++;
+            if ( begin == end )
+            {
+                break;
+            }
 
-    if(w->header_parse_tries > 1) {
-        if(w->header_parse_last_size < last_pos)
-            last_pos = 0;
-
-        if(strstr(&s[last_pos], "\r\n\r\n") == NULL) {
-            if(w->header_parse_tries > 10) {
-                info("Disabling slow client after %zu attempts to read the request (%zu bytes received)", w->header_parse_tries, buffer_strlen(w->response.data));
-                w->header_parse_tries = 0;
-                w->header_parse_last_size = 0;
-                web_client_disable_wait_receive(w);
-                return HTTP_VALIDATION_NOT_SUPPORTED;
+            if (*begin == '\n')
+            {
+                counter++;
             }
+        } else if (*begin == '\n') {
```

Comment:
Logic: same as above, what if we have strings like `"\n\nAB"`?

Reply:
This is worse. While we were developing @ilyam8 gave us a suggestion to move part of the code to the libnetdata, the suggestion was accept and this function was copied to there, but after the tests I forgot to remove it from here. Thanks by the review!